optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/uni/DroolsUniCondition.java

```diff
@@ -76,41 +81,69 @@ private DroolsUniCondition(DroolsMetadata<Object, A> aMetadata) {
-            return new DroolsBiCondition<>(aMetadata, ((DroolsInferredMetadata) bMetadata).substitute(newPattern));
-        } else {
-            return new DroolsBiCondition<>(aMetadata, ((DroolsGenuineMetadata) bMetadata).substitute(newPattern));
-        }
+        return new DroolsBiCondition<>(aMetadata, bMetadata.substitute(patternSupplier));
+    }
+
+    public <ResultContainer, NewA, NewB> List<RuleItemBuilder<?>> completeWithLogicalInsert(
+            Object ruleId, Function<A, NewA> groupKeyMapping,
+            UniConstraintCollector<A, ResultContainer, NewB> collector) {
+        Function1<Object, A> extractor = o -> getAMetadata().extract(o);
+        Function1<Object, NewA> grouper = a -> groupKeyMapping.apply(extractor.apply(a));
+        // Accumulate all NewA into a set.
+        Declaration<NewA> innerNewADeclaration = (Declaration<NewA>) declarationOf(Object.class);
+        PatternDSL.PatternDef<Object> innerNewACollectingPattern = getAMetadata().buildPattern()
+                .bind(innerNewADeclaration, grouper);
+        Declaration<Object> setOfNewADeclaration = declarationOf(Object.class);
+        ExprViewItem<Object> collectingPattern = DSL.accumulate(innerNewACollectingPattern,
+                accFunction(CollectSetAccumulateFunction::new, innerNewADeclaration).as(setOfNewADeclaration));
+        // Operate individually with every NewA from that set, creating the final grouping.
+        Declaration<NewA> actualNewADeclaration =
+                (Declaration<NewA>) declarationOf(Object.class, from(setOfNewADeclaration));
+        // And run an accumulate on all A which match NewA, applying the collector.
+        Predicate2<Object, NewA> matcher = (a, newA) -> grouper.apply(a).equals(newA);
+        Declaration<A> aVariable = (Declaration<A>) getAMetadata().getVariableDeclaration();
+        PatternDSL.PatternDef<Object> innerAccumulatePattern = aMetadata.buildPattern()
+                .expr(actualNewADeclaration, matcher)
+                .bind(aVariable, extractor);
+        AccumulateFunction<DroolsAccumulateContext<ResultContainer>> accumulateFunction =
+                new DroolsUniAccumulateFunctionBridge<>(collector);
+        Declaration<Object> accumulateResultVariable = declarationOf(Object.class);
+        ExprViewItem<Object> outerAccumulatePattern = DSL.accumulate(innerAccumulatePattern,
+                accFunction(() -> accumulateFunction, aVariable).as(accumulateResultVariable));
+        ConsequenceBuilder._2<?, ?> consequence = on(actualNewADeclaration, accumulateResultVariable)
+                .execute((drools, newA, newB) -> {
+                    RuleContext kcontext = (RuleContext) drools;
+                    kcontext.insertLogical(new DroolsLogicalTuple(ruleId, newA, newB));
+                });
+        return Arrays.asList(collectingPattern, pattern(setOfNewADeclaration), pattern(actualNewADeclaration),
+                outerAccumulatePattern, consequence);
     }
```

Comment:
This is the meat of the pull request. A single Drools rule that both groups and accumulates.

Reply:
when
	accumulate(Person(), $set: collectSet(Person::getCity))
        $newA : City() from $set // grouping
        accumulate(Person(getCity() == $newA), $newB: collector(ConstraintCollectors.count()))
then
	insertLogical($newA, $newB);
end